Chopsticks - adapted article to Tanssi (#78)
moondance-labs/tanssi-docs: builders/interact/substrate-api/chopsticks.md

---
title: How to use Chopsticks to Fork your Appchain
description: Learn the basics of how to use Chopsticks to replay blocks, dissect state changes, test XCM interactions, and create a local fork of your Tanssi ContainerChain.
---

# How to Use Chopsticks to Fork your Appchain

## Introduction {: #introduction }

> [Chopsticks](https://github.com/AcalaNetwork/chopsticks){target=_blank} provides a developer-friendly method of locally forking existing Substrate-based chains. It allows for the replaying of blocks to easily examine how transactions affect the state, the forking of multiple Appchains for XCM testing, and more. This empowers developers to test and experiment with their own custom blockchain configurations in a local development environment, without the need to deploy a live network.

Reviewer: ```suggestion
[Chopsticks](https://github.com/AcalaNetwork/chopsticks){target=\_blank} provides a developer-friendly method of locally forking existing [Substrate-based](/learn/framework/overview/){target=\_blank} chains. It allows for the replaying of blocks to easily examine how transactions affect the state, the forking of multiple Appchains for XCM testing, and more. This empowers developers to test and experiment with their own custom blockchain configurations in a local development environment, without the need to deploy a live network.
```

Overall, Chopsticks aims to simplify the process of building blockchain applications on Substrate and make it accessible to a wider range of developers.

> In this article, using Chopsticks to fork and interact with the local copy of a Tanssi ContainerChain will be covered.

Reviewer: ```suggestion
In this article, using Chopsticks to fork and interact with the local copy of a Tanssi ContainerChain will be covered.

!!! note
    Chopsticks currently does not support calls done via the Ethereum JSON-RPC. Consequently, you can't fork your chain using Chopsticks and connect Metamask to it.
```

## Prerequisites {: #prerequisites }

The original Chopsticks project was initially created to support the [Babe](https://docs.substrate.io/reference/glossary/#blind-assignment-of-blockchain-extension-babe){target=\_blank} and [Aura](https://docs.substrate.io/reference/glossary/#authority-round-aura){target=\_blank} block production algorithms. However, to ensure compatibility with the Tanssi block production as a service protocol, modifications have been made and published in a new [fork](https://github.com/moondance-labs/chopsticks.git){target=\_blank} of the original repository.

> To follow the steps of this article, the first step will be cloning the repository along with its submodules([smoldot](https://github.com/smol-dot/smoldot.git){target=\_blank}):

Reviewer: ```suggestion
Therefore, to use Chopsticks to fork your Appchain, the first step is to clone the fork repository along with its submodules ([Smoldot](https://github.com/smol-dot/smoldot.git){target=\_blank}):
```

```bash
 cd chopsticks && yarn
```

Finally, build the project:

```bash
yarn build-wasm
```

Now the development environment is ready to start testing and debugging Tanssi-deployed ContainerChains.

## Forking a Demo EVM ContainerChain with Chopsticks {: #forking-demo-chain }

To fork an Appchain using Chopsticks, simply execute the command with only the RPC endpoint as a parameter:

```bash
yarn start --endpoint wss://fraa-dancebox-3001-rpc.a.dancebox.tanssi.network
```

This will start a local clone of the chain as it was in the latest block.

--8<-- 'code/builders/interact/substrate-api/chopsticks/chopsticks-1.md'

Typically, the configuration parameters are included in a configuration file, such as those already included in the repository's `configs` folder for the relay chains and parachains deployed in the Dotsama ecosystem. The following configuration file works for the [Demo EVM ContainerChain](/builders/tanssi-network/networks/dancebox/demo-evm-containerchain/){target=\_blank}, overriding the chain's sudo account with Alith's and additionally funding the account with tokens:

> === "Demo EVM ContainerChain"

Author: We'll not have another config file, at least in the short term
[Tab removed]

    ```yaml
    endpoint: wss://fraa-dancebox-3001-rpc.a.dancebox.tanssi.network

>     mock-signature-host: true

Reviewer: ```suggestion
    mock-signature-host: true
    allow-unresolved-imports: true
```

    db: ./tmp/db_ftrcon.sqlite

    import-storage:
        System:
            Account:

>                 - - - "0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac" # Alith

Reviewer: Are you sure this works? Normally you need to respect tab padding, you see this `-  -  -` because of prettier linting your yaml file

Author: Yes, it works!

|   `mock-signature-host`    | Mock signature host so that any signature starts with `0xdeadbeef` and filled by `0xcd` is considered valid. |

You can run the command `yarn start` to fork chains by specifying a local configuration file. Alternatively, if the chain is listed in the repository's `configs` folder, either the name or the GitHub URL can be used. 

=== "Local File Path"

    ```bash
    yarn start --config=configs/polkadot.yml
    ```

=== "Chain Name"

    ```bash
    yarn start --config=polkadot
    ```

=== "GitHub URL"

    ```bash
    yarn start \
    --config=https://github.com/moondance-labs/chopsticks.git/master/configs/polkadot.yml
    ```

All of the settings (except `genesis` and `timestamp`) can also be passed as flags to configure the environment completely in the command line. For example, the following command forks the Demo EVM ContainerChain at block 100.

```bash
yarn start --endpoint wss://fraa-dancebox-3001-rpc.a.dancebox.tanssi.network --block 100
```

### Interacting with a Fork {: #interacting-with-a-fork }

When running a fork, by default it will be accessible at:

```text
ws://localhost:8000
```

> You will be able to interact with the parachain via libraries such as [Polkadot.js](https://github.com/polkadot-js/common){target=_blank} and its [user interface, Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A8000#/explorer){target=_blank}.  

Reviewer: ```suggestion
You will be able to interact with the parachain via libraries such as [Polkadot.js](https://github.com/polkadot-js/common){target=\_blank} and its [user interface, Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A8000#/explorer){target=\_blank}.  
```

> You can interact with Chopsticks via the [Polkadot.js Apps hosted user interface](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A8000#/explorer){target=_blank}:

Reviewer: ```suggestion
You can interact with Chopsticks via the [Polkadot.js Apps hosted user interface](https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A8000#/explorer){target=\_blank}:
```

![Polkadot Js](/images/builders/interact/substrate-api/chopsticks/chopsticks-1.png)

You should now be able to interact with the forked chain as you would with the real one.

!!! note

>     If your browser cannot connect to the WebSocket endpoint provided by Chopsticks, you might need to allow insecure connections for the Polkadot.js Apps URL. Another solution is to run the [Docker version of Polkadot.js Apps](https://github.com/polkadot-js/apps#docker){target=_blank}.

Reviewer: ```suggestion
    If your browser cannot connect to the WebSocket endpoint provided by Chopsticks, you might need to allow insecure connections for the Polkadot.js Apps URL. Another solution is to run the [Docker version of Polkadot.js Apps](https://github.com/polkadot-js/apps#docker){target=\_blank}.
```

|      `wasm-override`       |  Path of the WASM to use as the parachain runtime, instead of an endpoint's runtime.   |
|            `db`            |    Path to the name of the file that stores or will store the parachain's database.    |
|          `config`          |                            Path or URL of the config file.                             |
| `output-path=/[file_path]` |   Use to print out results to a JSON file instead of printing it out in the console.   |
|           `html`           | Include to generate an HTML representation of the storage diff preview between blocks. |
|           `open`           |                        Whether to open the HTML representation.                        |

For example, running the following command will re-run the Demo EVM ContainerChains's block 1000, and write the storage diff and other data in a `chain-output.json` file:  

```bash
yarn start run-block  \
--endpoint wss://fraa-dancebox-3001-rpc.a.dancebox.tanssi.network  \
--output-path=./chain-output.json  \
--block 1000
```

## WebSocket Commands {: #websocket-commands }

> Chopsticks' internal websocket server has special endpoints that allow the manipulation of the local Substrate chain. These are the methods that can be invoked:  

Reviewer: ```suggestion
Chopstick's internal websocket server has special endpoints that allow the manipulation of the local Substrate chain. These are the methods that can be invoked:  
```

> |      Method      |      Parameters       |                          Description                          |

Author: Better now? wdyt?

|     `date`     |               `Date`                |                        `"2030-08-15T00:00:00"`                         |
| `hashOrNumber` |               `number               |                                string`                                 |

- **`options` { "to": number, "count": number }** - a JSON object where `"to"` will create blocks up to a certain value, and `"count"` will increase by a certain number of blocks. Use only one entry at a time within the JSON object  
- **`values` Object** - a JSON object resembling the path to a storage value, similar to what you would retrieve via Polkadot.js  
- **`blockHash` string** - optional, the blockhash at which the storage value is changed  
- **`date` Date** - a Date string (compatible with the JavaScript Date library) that will change the time stamp from which the next blocks being created will be at. All future blocks will be sequentially after that point in time  
- **`hashOrNumber` number | string** - if found, the chain head will be set to the block with the block number or block hash of this value  

> Each method can be invoked by connecting to the websocket (`ws://localhost:8000` by default) and sending the data and parameters in the following format. Replace `METHOD_NAME` with the name of the method, and replace or delete `PARAMETER_1` and `PARAMETER_2` with the parameter data relevant to the method:  

Reviewer 2: ```suggestion
Each method can be invoked by connecting to the WebSocket (`ws://localhost:8000` by default) and sending the data and parameters in the following format. Replace `METHOD_NAME` with the method's name, and replace or delete `PARAMETER_1` and `PARAMETER_2` with the parameter data relevant to the method:  
```

```bash
yarn start run-block  \
--endpoint wss://fraa-dancebox-3001-rpc.a.dancebox.tanssi.network  \
--output-path=./chain-output.json  \
--block 1000
```

## WebSocket Commands {: #websocket-commands }

Chopstick's internal WebSocket server has special endpoints that allow the manipulation of the local Substrate chain. 

These are the methods that can be invoked and their parameters:

|      Method      |      Parameters       |                          Description                          |
|:----------------:|:---------------------:|:-------------------------------------------------------------:|
|  `dev_newBlock`  |       `options`       |               Generates one or more new blocks.               |
| `dev_setStorage` | `values`, `blockHash` |         Create or overwrite the value of any storage.         |
| `dev_timeTravel` |        `date`         |     Sets the timestamp of the block to the `date` value.      |
|  `dev_setHead`   |    `hashOrNumber`     | Sets the head of the blockchain to a specific hash or number. |

> ??? code "Parameters details"

Reviewer 2: I love a good expandable code snippet moment! haha

Reviewer 2: After continuing my review, I feel like the initial problem wasn't addressed. So if you want to implement an expandable snippet for each item (like we have in [Moonbeam docs](https://docs.moonbeam.network/builders/interoperability/xcm/xc20/send-xc20s/xtokens-pallet/#extrinsics)), you can. Or if there aren't too many tabbed items spamming the screen, you don't have to use expandable snippets

    |  `blockHash`      |              `string`               | `"0x1a34506b33e918a0106b100db027425a83681e2332fe311ee99d6156d2a91697"` |
    |     `date`        |               `Date`                |                        `"2030-08-15T00:00:00"`                         |
    | `hashOrNumber`    |         `number | string`           |                                                                        |

    - **`options` { "to": number, "count": number }** - a JSON object where `"to"` will create blocks up to a certain value, and `"count"` will increase by a certain number of blocks. Use only one entry at a time within the JSON object  
    - **`values` Object** - a JSON object resembling the path to a storage value, similar to what you would retrieve via Polkadot.js  
    - **`blockHash` string** - optional, the block hash at which the storage value is changed  
    - **`date` Date** - a Date string (compatible with the JavaScript Date library) that will change the timestamp at which the next blocks being created will be. All future blocks will be sequentially created after that point in time  

>     - **`hashOrNumber` number | string** - if found, the chain head will be set to the block with the block number or block hash of this value  

Reviewer 2: I think what I don't like about this is that we have the parameter information for the same parameter in two different spots. Like we have the parameter format and example in the table then below the table there is a description for that same parameter.

I would say let's just add a description column to the table so that the parameter details are all in one place, but I think that would make the table look too crowded and it wouldn't be the most readable.

So with all of this in mind, and the fact that these parameters are reused across the different methods, I think we should do something like this:

```
- `dev_newBlock` - generates one or more new blocks

     === "Parameters"
         
         - `options` -  a JSON object where `"to"` will create blocks up to a certain value, and `"count"` will increase by a certain number of blocks. Use only one entry at a time within the JSON object. This can be formatted as follows: `{ "to": number, "count": number }`
         
     === "Example"
     
         ```js
         import { WsProvider } from '@polkadot/api'
         const ws = new WsProvider(`ws://localhost:8000`)
         await ws.send('dev_newBlock', [{ count: 5 }])
         ```
```

The example is probably not formatted correctly to be able to copy/paste in a terminal, because I just wanted to get my general idea across. But I think if we can clean it up like this, it will look a lot nicer and be easier to read.

If you have questions, just ping me in slack 🙂

Author: Marking it as solved, even though it's wip in another branch.
